// testcafe/lib/reporter/command-report-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const observation_1 = require("../test-run/commands/observation");
const actions_1 = require("../test-run/commands/actions");
class CommandReportItem {
    constructor(command) {
        this.type = command.type;
        if (command instanceof observation_1.ExecuteSelectorCommand)
            this.selector = this._prepareSelector(command);
        else if (command instanceof observation_1.ExecuteClientFunctionCommand)
            this.clientFn = this._prepareClientFunction(command);
        else if (command instanceof actions_1.UseRoleCommand)
            this.role = this._prepareRole(command);
        else if (command instanceof actions_1.NavigateToCommand)
            this.url = this._prepareUrl(command);
        else if (command instanceof actions_1.SetNativeDialogHandlerCommand)
            this.dialogHandler = this._prepareDialogHandler(command);
        else
            this._prepareProperties(command);
    }
    _prepareSelector(selector) {
        const selectorChain = selector.apiFnChain;
        return selectorChain.join('');
    }
    _prepareClientFunction(fn) {
        return {
            code: fn.fnCode,
            args: fn.args[0]
        };
    }
    _prepareDialogHandler(command) {
        return this._prepareClientFunction(command.dialogHandler);
    }
    _prepareRole(roleCommand) {
        const { loginPage, opts, phase } = roleCommand.role;
        return { loginPage, options: opts, phase };
    }
    _prepareUrl(navigateCommand) {
        return navigateCommand.url;
    }
    _prepareProperties(command) {
        if (!command._getAssignableProperties)
            return;
        const sourceProperties = command._getAssignableProperties().map(prop => prop.name);
        sourceProperties.forEach((key) => {
            const prop = command[key];
            if (prop instanceof observation_1.ExecuteSelectorCommand)
                this[key] = this._prepareSelector(prop);
            else
                this[key] = prop;
        });
    }
}
exports.CommandReportItem = CommandReportItem;
